codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,66 +246,43 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
-
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
+            ChannelFuture future;
+            if (openStream) {
+                // Write the initial headers on the stream immediately
+                stream.open(endOfStream);
+                future = frameWriter.writeHeaders(ctx, stream.id(), headers, streamDependency, weight,
+                        exclusive, padding, endOfStream, promise);
+                ctx.flush();
+            } else {
+                // Pass intermediate & trailing headers to the flow-controller so it can maintain their
+                // sequence relative to DATA frames.
+                future = flowController().sendFlowControlledData(ctx, stream,
+                        new FlowControlledHeaders(ctx, headers, padding, endOfStream, streamId, promise),
+                        promise);
+            }
+            if (endOfStream) {
+                // Flag delivery of EOS synchronously to prevent subsequent frames being enqueued in the flow
+                // controller.
+                stream.endOfStreamSent();
             }
+            future.addListener(new ChannelFutureListener() {
```

Comment:
Do we have to do this for every frame?  I think the previous code was only doing this for a frame which was not sent "immediately".  Perhaps this is only needed if we are calling the flow controller to queue up the frame?